[16.0][ADD] stock_release_channel_shipment_advice_deliver

jbaudoux opened the pull request:
This module adds an action to the release channel to automate the delivery of its shippings through shipment advices.

Depends on:
* https://github.com/OCA/stock-logistics-transport/pull/114

@sbejaoui I refactored a bit the code to drop the dependency on:
* stock_picking_started : stock_available_to_promise_release already does the check on printed
* on propagate release channel : picking_chain_ids is more accurate https://github.com/OCA/wms/blob/16.0/stock_release_channel/models/stock_release_channel.py#L125
* process end date: I could not figure out why this was a dependency
* stock_picking_backorder_reason needs to be updated in the button_validate

cc @sebalix @simahawk @lmignon

jbaudoux committed:
[ADD] stock_release_channel_shipment_advice_deliver

jbaudoux commented:
TODO: rebuild fr translation

sbejaoui commented on stock_release_channel_deliver/models/shipment_advice.py:
```diff
@@ -0,0 +1,108 @@
+    def _get_picking_not_to_backorder(self):
+        pickings_with_backorder = self.planned_picking_ids._check_backorder()
+        picking_not_to_backorder = pickings_with_backorder.filtered(
+            lambda p: p.partner_id.sale_reason_backorder_strategy == "cancel"
```
this depends to `stock_picking_backorder_reason`, we may need a glue module

jbaudoux replied:
I have the impression this can be integrated in `stock_picking_backorder_reason` in the `button_validate` method.

sbejaoui commented on stock_release_channel_deliver/models/stock_picking.py:
```diff
@@ -0,0 +1,14 @@
+
+
+class StockPicking(models.Model):
+    inherit = "stock.picking"
```
typo 

```suggestion
    _inherit = "stock.picking"
```

jbaudoux commented:
> TODO: rebuild fr translation

done

jbaudoux commented on stock_release_channel_deliver/models/shipment_advice.py:
```diff
@@ -0,0 +1,108 @@
+                    self.action_in_progress()
+                picking_not_to_backorder = self._get_picking_not_to_backorder()
+                self.with_context(
+                    picking_ids_not_to_backorder=picking_not_to_backorder.ids
```
@sbejaoui I can't find where this context variable `picking_ids_not_to_backorder` is used. Can you clarify?

jbaudoux replied:
ah spotted in standard odoo ;)

tuantrantg committed:
[IMP] stock_release_channel_shipment_advice_deliver

Add option auto_delivery and allow to print in-process shipment advice

tuantrantg committed:
[IMP] stock_release_channel_shipment_advice_deliver

Allow to print shipment in delivered state

jbaudoux committed:
[IMP] stock_release_channel_shipment_advice_deliver

Allow to deliver when the picking is partially processed but no backorder is created

jbaudoux commented:
Added support for partial picking without backorder. The `aciton_deliver` doesn't raise an error anymore that it cannot be unreleased. See 948637a525db6afd193ee365e40ba4fc13a5a16c

jbaudoux committed:
[FIX] stock_release_channel_shipment_advice_deliver

Fix test as channels are now created by default in state asleep

jbaudoux committed:
[IMP] stock_release_channel_shipment_advice_deliver

Name the tests

jbaudoux commented:
@sebalix @sbejaoui @lmignon @rousseldenis Can you update / put review on this one ?

jbaudoux reviewed (commented):
Planned upcoming improvements:
* Add the action "Deliver" (already existing on the form view) on the kanban view (in the 3 dots)
* If "Auto Deliver", hide "Plan Shipment" action
* If we create manually a Shipment Advice from a release channel, when we start it, we have to set in 'delivering' state its release channel

tuantrantg committed:
[IMP] stock_release_channel_shipment_advice_deliver

- Show button 'Deliver' in the kanban view
- Show button 'Plan Shipment' if not auto deliver
- Update state of release channel to 'delivering' when the related shipment advice starts

tuantrantg commented:
> Planned upcoming improvements:
> 
> * Add the action "Deliver" (already existing on the form view) on the kanban view (in the 3 dots)
> * If "Auto Deliver", hide "Plan Shipment" action
> * If we create manually a Shipment Advice from a release channel, when we start it, we have to set in 'delivering' state its release channel

hi @jbaudoux [here](https://github.com/jbaudoux/wms/pull/6) is the PR for these improvements, please take a look

rousseldenis reviewed (approved):
Code review

jbaudoux commented on stock_release_channel_shipment_advice_deliver/models/shipment_advice.py:
```diff
@@ -0,0 +1,106 @@
+
+    def action_in_progress(self):
+        res = super().action_in_progress()
+        self.release_channel_id.filtered(
```
We need a test for this last improvement

tuantrantg replied:
Hi @jbaudoux I have a trouble about this
* I create a picking (called P1) with a release channel, called C1 (C1's state is `open`, planning method is `simple`)
* I also create a shipment advice (called SA1) from C1 -> `picking_to_plan_ids` contains P1
* After that I plan picking P1 in shipment SA1 -> `picking_to_plan_ids` will be null because picking is planned
* I lock the SA1 and check auto_deliver
* If I continue to start the SA1 -> nothing happens in C1 because `is_action_deliver_allowed` is calculated based on `picking_to_plan_ids` but it is null now ([here](https://github.com/OCA/wms/pull/819/files#diff-4d3327d6e5d2a44ca6373fd21fd676ce1f565a0c691f73b5b3b09bde0950152fR59)), so C1 cannot change the state

Could you please give me an advice in this one? Maybe I've missed something

jbaudoux replied:
* When you create the SA1, make it locked and auto_deliver
* You should skip the manual planning of P1. You should not be able to do manual planning in auto_deliver

tuantrantg replied:
Hi @jbaudoux the test is [here](https://github.com/jbaudoux/wms/pull/7) please take a look

rousseldenis replied:
@tuantrantg Is this solved ?

jbaudoux commented:
@sbejaoui Can you update your review on this one?

cyrilmanuel commented:
Hello @sbejaoui, gentle reminder, can you update the review please :)

jbaudoux commented:
/ocabot merge nobump

rousseldenis committed:
[FIX] stock_release_channel_shipment_advice_deliver: allow to deliver with a picking backorder

As we want to let the operator deliver even if a backorder still exists,
show the confirmation wizard in that case (and not an error).

rousseldenis committed:
[IMP] stock_release_channel_shipment_advice_deliver: Be more lax on backorder pickings

Don't raise error if an unstarted backorder still remains in the chain

rousseldenis commented:
/ocabot merge nobump

rousseldenis commented:
@jbaudoux Could you rebase?

jbaudoux committed:
stock_release_channel_shipment_advice_deliver: fix notification

Test if the shipment advice is linked to a release channel for sending notification

jbaudoux committed:
[IMP] stock_release_channel

Define open_picking_ids as picking_ids is growing over time

jbaudoux committed:
[PERF] stock_release_channel_shipment_advice_deliver

Limit loaded channel picking_ids

jbaudoux commented:
@rousseldenis I cherry-picked your improvements. I reverted one for what I think a simpler solution. Can you check it? Then I'll squash the commit & revert

jbaudoux committed:
stock_available_to_promise_release: minor fixes

jbaudoux committed:
[I18N] stock_release_channel_shipment_advice_deliver: fr

jbaudoux committed:
stock_release_channel: minor fix

jbaudoux commented:
> TODO: rebuild fr translation

done

jbaudoux commented:
/ocabot merge minor
